### PR TITLE
Behavior Discovered Failure for PlaceOntop

### PR DESCRIPTION
**Pull request by wmcclinton**
This PR adds a Discovered Failure Exception in the PlaceOntop option model. This occurs when we try to place an object on top of another and end up placing it on a different object. We assume that object is blocking it and throw a Discovered Failure Exception with that object.

### CODE REVIEW

File: predicators/behavior_utils/option_model_fns.py [hunk 1]
```diff
@@ -145,6 +151,54 @@ def placeOntopObjectOptionModel(_init_state: State,
         for _ in range(15):
             env.step(np.zeros(env.action_space.shape))
 
+        ####
+        # Check whether object is ontop of not a target object
+        from igibson.object_states.adjacency import VerticalAdjacency
+        from igibson.object_states.touching import Touching
```
Comment (NishanthJKumar) [inline predicators/behavior_utils/option_model_fns.py]: Can we do this import at the top?

File: predicators/behavior_utils/option_model_fns.py [hunk 2]
```diff
@@ -145,6 +151,54 @@ def placeOntopObjectOptionModel(_init_state: State,
...
+                            # but for the moment, we just need position and orientation.
+                            obj_type = Type(
+                                type_name,
+                                [
+                                    "pos_x", "pos_y", "pos_z", "orn_0",
+                                    "orn_1", "orn_2", "orn_3"
+                                ],
+                            )
+                            _type_name_to_type[type_name] = obj_type
+                        #
```
Comment (NishanthJKumar) [inline predicators/behavior_utils/option_model_fns.py]: unnecessary comment, remove?

File: predicators/behavior_utils/option_model_fns.py [hunk 3]
```diff
@@ -145,6 +151,54 @@ def placeOntopObjectOptionModel(_init_state: State,
...
+                        if isinstance(obj, (URDFObject, RoomFloor)):
+                            if "board_game" in obj.name:
+                                obj_name = obj.name + ".n.01_1"
+                            else:
+                                obj_name = obj.bddl_object_scope
+                        else:
+                            assert isinstance(obj, BRBody)
+                            obj_name = "agent"
+                        #
```
Comment (NishanthJKumar) [inline predicators/behavior_utils/option_model_fns.py]: ditto

File: predicators/behavior_utils/option_model_fns.py [hunk 4]
```diff
@@ -145,6 +151,54 @@ def placeOntopObjectOptionModel(_init_state: State,
...
+        objs_under = set()
+        for obj in env.scene.get_objects():
+            if not obj_in_hand.states[Touching].get_value(obj):
+                continue
+            adjacency = obj.states[VerticalAdjacency].get_value()
+            if obj_in_hand.get_body_id() in adjacency.positive_neighbors:
+                if "floor" not in obj.category:
+                    if obj != obj_to_place:
+                        type_name = obj.category
+                        #
```
Comment (NishanthJKumar) [inline predicators/behavior_utils/option_model_fns.py]: unnecessary comment, remove?

File: predicators/behavior_utils/option_model_fns.py [hunk 5]
```diff
@@ -145,6 +151,54 @@ def placeOntopObjectOptionModel(_init_state: State,
...
+                        _type_name_to_type: Dict[str, Type] = {}
+                        for type_name in ALL_RELEVANT_OBJECT_TYPES:
+                            # type_name = ig_obj.category
```
Comment (NishanthJKumar) [inline predicators/behavior_utils/option_model_fns.py]: unnecessary comment?

File: predicators/envs/behavior.py [hunk 6]
```diff
@@ -289,6 +289,9 @@ def _get_tasks(self,
                             while self.task_instance_id in broken_instances[
                                     'train']:
                                 self.task_instance_id = rng.integers(0, 10)
+                # ### SET TASK IDs
+                self.task_instance_id = 13
+                # ###
```
Comment (NishanthJKumar) [inline predicators/envs/behavior.py]: revert this change!

File: predicators/envs/behavior.py [hunk 7]
```diff
@@ -508,7 +508,12 @@ def render_state(self,
                         "behavior_mode in settings.py instead")
 
     def _get_task_relevant_objects(self) -> List["ArticulatedObject"]:
-        return list(self.igibson_behavior_env.task.object_scope.values())
+        additional_objs = [
+            obj for obj in self.igibson_behavior_env.scene.get_objects()
+            if "board_game" in obj.name
+        ]
```
Comment (NishanthJKumar) [inline predicators/envs/behavior.py:518]: leave a comment explaining this so that we don't forget in the future/new ppl understand what's going on here

File: predicators/envs/behavior.py [hunk 8]
```diff
@@ -756,6 +761,8 @@ def _closed_classifier(self, state: State, objs: Sequence[Object]) -> bool:
     @staticmethod
     def _ig_object_name(ig_obj: "ArticulatedObject") -> str:
         if isinstance(ig_obj, (URDFObject, RoomFloor)):
+            if "board_game" in ig_obj.name:
+                return ig_obj.name + ".n.01_1"
```
Comment (NishanthJKumar) [inline predicators/envs/behavior.py:769]: i'm so confused; how are board_game objects of type RoomFloor?

**Review by NishanthJKumar — COMMENTED** [review]
Left some things to explain/fix, but mostly looks good!

File: predicators/envs/behavior.py (hunk 7 above)
Comment (wmcclinton) [reply predicators/envs/behavior.py:518]: Good point.

File: predicators/envs/behavior.py (hunk 8 above)
Comment (wmcclinton) [reply predicators/envs/behavior.py:769]: That is an isinstance of URDFObject or RoomFloor.

**Review by NishanthJKumar — APPROVED** [review]
LGTM!
